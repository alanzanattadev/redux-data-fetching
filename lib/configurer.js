"use babel";
// @flow
import { schema } from "normalizr";
import { buildSchema } from "graphql";
import type { GraphQLSchema } from "graphql";
import configureMiddleware from "./middleware";
import configureReducer from "./reducer";
import configureActions from "./actions";
import configureConnecter from "./hoc";
import { graphQLizr, graphQLRecordr } from "./graphqlTypesConverters";
import { convertsTypesSchemaToSelectorSchema } from "./selectors";

export function configure(
  graphQLSchema: GraphQLSchema | string,
  context: mixed,
  { __debug = false }: { __debug?: boolean } = {},
) {
  const graphQLCompiledSchema =
    typeof graphQLSchema === "string"
      ? buildSchema(graphQLSchema)
      : graphQLSchema;
  const selectorSchema = convertsTypesSchemaToSelectorSchema(
    graphQLCompiledSchema,
    { __debug },
  );
  const normalizrModel = graphQLizr(graphQLCompiledSchema);
  const recordsModel = graphQLRecordr(graphQLCompiledSchema);
  const actions = configureActions();
  const middleware = configureMiddleware(
    graphQLCompiledSchema,
    actions,
    normalizrModel,
    context,
  );
  const reducer = configureReducer(
    normalizrModel.entities,
    recordsModel,
    graphQLCompiledSchema,
  );
  const { GraphQLConnecter, DataFetcher, DataHandlers } = configureConnecter({
    selectorSchema,
    typesSchema: graphQLCompiledSchema,
    recordTypes: recordsModel,
    actions,
  });
  return {
    actions,
    middleware,
    reducer,
    normalizrModel,
    recordsModel,
    selectorSchema,
    GraphQLConnecter,
    DataFetcher,
    DataHandlers,
  };
}
