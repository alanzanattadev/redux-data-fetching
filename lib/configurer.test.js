import {
  GraphQLSchema,
  GraphQLObjectType,
  GraphQLList,
  GraphQLString,
  buildSchema,
} from "graphql";
import { configure } from "./configurer";

describe("Configurer", () => {
  it("returns the differents models", () => {
    const subject = configure(
      new GraphQLSchema({
        query: new GraphQLObjectType({
          name: "RootQuery",
          fields: {
            users: {
              type: new GraphQLObjectType({
                name: "User",
                fields: {
                  id: { type: GraphQLString },
                },
              }),
            },
          },
        }),
      }),
    );

    expect(subject).toMatchSnapshot();
  });

  it("works with schema as string", () => {
    const subject = configure(
      `
      type User {
        id: String!,
        name: String,
        friends: [User!]
      }
      
      type Query {
        users: [User!],
        user(id: String): User
      }
      
      type Mutation {
        createUser(name: String!): User
      }
    `,
      {},
      {
        users: () => [
          { id: "user1", name: "Alan" },
          { id: "user2", name: "Antoine" },
        ],
        user: () => ({ id: "user1", name: "Alan", friends: [] }),
        createUser: ({ name }) => ({ id: "id", name: name }),
      },
    );

    expect(subject).toMatchSnapshot();
  });
});
