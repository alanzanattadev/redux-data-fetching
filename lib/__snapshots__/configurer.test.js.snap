// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configurer returns the differents models 1`] = `
Object {
  "DataFetcher": [Function],
  "DataHandlers": [Function],
  "GraphQLConnecter": [Function],
  "actions": Object {
    "bustQueryCache": [Function],
    "fetchData": [Function],
    "mutateData": [Function],
    "packageData": [Function],
    "queryFailed": [Function],
    "queryStarted": [Function],
    "removeData": [Function],
  },
  "middleware": [Function],
  "normalizrModel": Object {
    "converters": Object {
      "users": "User",
    },
    "entities": Object {
      "User": EntitySchema {
        "_getId": [Function],
        "_idAttribute": "id",
        "_key": "User",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
    },
  },
  "recordsModel": Object {
    "RootQuery": [Function],
    "User": [Function],
  },
  "reducer": [Function],
  "selectorSchema": GraphQLSchema {
    "_directives": Array [
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": undefined,
            "description": "Included when true.",
            "name": "if",
            "type": "Boolean!",
          },
        ],
        "astNode": undefined,
        "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "include",
      },
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": undefined,
            "description": "Skipped when true.",
            "name": "if",
            "type": "Boolean!",
          },
        ],
        "astNode": undefined,
        "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "skip",
      },
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": "No longer supported",
            "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted in [Markdown](https://daringfireball.net/projects/markdown/).",
            "name": "reason",
            "type": "String",
          },
        ],
        "astNode": undefined,
        "description": "Marks an element of a GraphQL schema as no longer supported.",
        "locations": Array [
          "FIELD_DEFINITION",
          "ENUM_VALUE",
        ],
        "name": "deprecated",
      },
    ],
    "_implementations": Object {},
    "_mutationType": undefined,
    "_queryType": "SelectorsRootQueryType",
    "_subscriptionType": undefined,
    "_typeMap": Object {
      "Boolean": "Boolean",
      "SelectorsRootQueryType": "SelectorsRootQueryType",
      "String": "String",
      "UserSelector": "UserSelector",
      "__Directive": "__Directive",
      "__DirectiveLocation": "__DirectiveLocation",
      "__EnumValue": "__EnumValue",
      "__Field": "__Field",
      "__InputValue": "__InputValue",
      "__Schema": "__Schema",
      "__Type": "__Type",
      "__TypeKind": "__TypeKind",
    },
    "astNode": null,
  },
}
`;
